pydantic_settings/sources.py

```diff
@@ -247,24 +249,74 @@ def __call__(self) -> dict[str, Any]:
         pass
 
 
+class DefaultSettingsSource(PydanticBaseSettingsSource):
+    """
+    Source class for loading default object values.
+    """
+
+    def __init__(self, settings_cls: type[BaseSettings], default_objects_copy_by_value: bool | None = None):
+        super().__init__(settings_cls)
+        self.defaults: dict[str, Any] = {}
+        self.default_objects_copy_by_value = (
+            default_objects_copy_by_value
+            if default_objects_copy_by_value is not None
+            else self.config.get('default_objects_copy_by_value', False)
+        )
+        for field_name, field_info in settings_cls.model_fields.items():
+            if not self.default_objects_copy_by_value:
+                if is_dataclass(type(field_info.default)) or is_model_class(type(field_info.default)):
+                    self.defaults[_field_name_for_signature(field_name, field_info)] = field_info.default
```

Comment:
@hramezani also consider whether this case should be kept or removed (lines 267-268)

Currently, without these lines, the below behavior is observed:

```python
class SubModel(BaseModel):
    val: int = 0
    flag: bool = False

ORIGINAL_OBJECT = SubModel()

class SettingsCopyByReference(BaseSettings):
    model_config = SettingsConfigDict(env_nested_delimiter='__')

    default_object: SubModel = ORIGINAL_OBJECT


by_ref = SettingsCopyByReference(default_object=ORIGINAL_OBJECT)
assert by_ref.default_object is ORIGINAL_OBJECT # Passes for default object in kwargs

try:
    by_ref = SettingsCopyByReference()
    assert by_ref.default_object is ORIGINAL_OBJECT, 'Fails for default object in class'
except AssertionError as err:
    print(err)
    #> Fails for default object in class
```

My assumption is the behavior should be the same. Therefore, if we keep these lines it **_fixes_** the above such that behavior is the same for both class and kwarg default objects.

If not, we can remove these lines and keep current behavior.

Reply:
Agreed. Also, `BaseModel` does not return the same instance for class default objects.

Should we change the flag name to something else more appropriate? I'm not sure what to call it, do you have any thoughts?

Reply:
The flag name should refer to partial updates and nested objects. like `nested_model_partial_update`.

I am open to your suggestions as well.